platforms/jvm/scala/src/main/java/org/gradle/language/scala/tasks/AbstractScalaCompile.java

```diff
@@ -236,6 +245,21 @@ private Map<File, File> resolveAnalysisMappingsForOtherProjects() {
                 }
             }
         }
+
+        return analysisMap;
+    }
+
+    private Map<File, File> resolveAnalysisMappingsForLocalSourceSets() {
```

Comment:
❓ The test is a good reproducer, but this won't work for a few reasons:
- We don't want tasks reaching across projects or into other tasks to get extra information.
- We want dependencies (like on incremental analysis) to be visible through other tools (like dependencies report or build scans)
- We're not always going to be able to get the analysis file (e.g., for non-Scala languages)

I looked at this a bit today and I think the problem is deeper. My expectation would be that if the incremental analysis file is missing for a given dependency, the compiler should fallback to recompiling more often or some other analysis, which doesn't seem to be happening. 

I found that this can be reproduced across projects and not just inside one project. If there's a directory on the classpath, changes to that directory are ignored unless we also have an incremental analysis file. I can't tell if this is expected Zinc behavior or if we're doing something unexpected. 

I was able to reproduce the problem with...
1. The provided test case
2. A Scala project that depends on the test fixtures of another Scala project. This passes something other than the main classes directory for the first Scala project.
3. A Scala project that depends on the test fixtures of another _Java_ project This passes something other than the main classes directory for the Java project.

With (3), that means just finding the analysis file isn't enough. We need to either change the way Zinc treats directories or we need to disallow directories as inputs. 

When a Scala project depends on a jar from another project, the problem doesn't occur because the jar itself changes between compiles. 

I also found if I take the given reproducer and add the project's jar as a dependency, this "fixes" the issue, but I think this is just tricking Zinc because it sees the jar changing. You can "fix" this by adding a dependency to the current project (`testImplementation(project(":"))`) because this adds the jar _and_ its analysis file. 

I suspect if there is a problem in our integration, it's going to be in ZincScalaCompiler or nearby. Debugging this is very painful because of the layers of processes (Test worker -> Gradle daemon -> Gradle compiler worker).

Does any of this ring a bell?

Reply:
Thanks for taking your time to thoroughly analyse the issue I observed.

> I found that this can be reproduced across projects and not just inside one project. If there's a directory on the classpath, changes to that directory are ignored unless we also have an incremental analysis file.

Yes. It seems that any change that happens outside of a simple `class` (like in companion `object`s) "semi-externally" to the current compilation requires that the previous analysis files are provided.

> I can't tell if this is expected Zinc behavior or if we're doing something unexpected.

My experience in both Gradle and Zinc is definitely lower than yours so I cannot tell it either ;-)

To sum it up...
Just as I thought, I came with an oversimplified view of the problem and by extension a "solution" to it ;-)
It seems that we should close the PR as it is more of a hack than fix.
Maybe I can strip it from the "solution" part and keep it as a reproducer attached to the issue.

Reply:
Oh, and thanks for an interesting workaround with tests depending on the project.
I will test if we may use it in the meantime without noticeable drawbacks.